packetbeat/protos/cassandra/parser.go

```diff
@@ -0,0 +1,246 @@
+package cassandra
+
+import (
+	"errors"
+	"time"
+
+	"fmt"
+	"github.com/elastic/beats/libbeat/common/streambuf"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/packetbeat/protos/applayer"
+	. "github.com/elastic/beats/packetbeat/protos/cassandra/internal/gocql"
+)
+
+type parser struct {
+	buf       streambuf.Buffer
+	config    *parserConfig
+	framer    *Framer
+	message   *message
+	onMessage func(m *message) error
+}
+
+type parserConfig struct {
+	maxBytes   int
+	compressor Compressor
+	ignoredOps map[string]interface{}
+}
+
+type message struct {
+	applayer.Message
+
+	// indicator for parsed message being complete or requires more messages
+	// (if false) to be merged to generate full message.
+	isComplete bool
+
+	failed  bool
+	ignored bool
+
+	data   map[string]interface{}
+	header map[string]interface{}
+	// list element use by 'transactions' for correlation
+	next *message
+
+	transactionTimeout time.Duration
+
+	results transactions
+}
+
+// Error code if stream exceeds max allowed size on append.
+var (
+	ErrStreamTooLarge = errors.New("Stream data too large")
+	isDebug           = false
+)
+
+func (p *parser) init(
+	cfg *parserConfig,
+	onMessage func(*message) error,
+) {
+	*p = parser{
+		buf:       streambuf.Buffer{},
+		config:    cfg,
+		onMessage: onMessage,
+	}
+
+	isDebug = logp.IsDebug("cassandra")
+
+}
+
+func (p *parser) append(data []byte) error {
+	_, err := p.buf.Write(data)
+	if err != nil {
+		return err
+	}
+
+	if p.config.maxBytes > 0 && p.buf.Total() > p.config.maxBytes {
+		return ErrStreamTooLarge
+	}
+	return nil
+}
+
+func (p *parser) feed(ts time.Time, data []byte) error {
+	if err := p.append(data); err != nil {
+		return err
+	}
+
+	for p.buf.Total() > 0 {
+		if p.message == nil {
+			// allocate new message object to be used by parser with current timestamp
+			p.message = p.newMessage(ts)
+		}
+
+		msg, err := p.parse()
+		if err != nil {
+			return err
+		}
+		if msg == nil {
+			break // wait for more data
+		}
+
+		// reset buffer and message -> handle next message in buffer
+		p.buf.Reset()
+		p.message = nil
+
+		// call message handler callback
+		if err := p.onMessage(msg); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *parser) newMessage(ts time.Time) *message {
+	return &message{
+		Message: applayer.Message{
+			Ts: ts,
+		},
+	}
+}
+
+func (p *parser) parse() (*message, error) {
+
+	// if p.frame is nil then create a new framer, or continue to process the last message
+	if p.framer == nil {
+		if isDebug {
+			logp.Debug("cassandra", "start new framer")
```

Comment:
use `debugf(format-string, ...)` instead of `logp.Debug("cassandra", format-string, ...)`. This reduces the chance of typos in debug selectors + reduces line length + some linters will check the parameters being correct if debugf is used.